functions/src/modules/gmail.functions.ts

```diff
@@ -0,0 +1,273 @@
+      return res.status(400).send('Invalid emailAddress.');
+    }
+
+    // Initiate gmail client
+    const { client_key, secret, redirect } = functions.config().gmail;
+    const oauth2Client = new Google.google.auth.OAuth2(
+      client_key,
+      secret,
+      redirect
+    );
+
+    const { oauth, emailAddress } = await fetchToken(email);
+
+    oauth2Client.setCredentials({
+      access_token: oauth.token,
+      refresh_token: oauth.refreshToken,
+    });
+    const gm = await Google.google.gmail({
+      version: 'v1',
+      auth: oauth2Client,
+    });
+
+    try {
+      // Make sure we have the correct labelId because GMail internally gives a random one that is
+      // not based on the given name, so we need to filter it by the given name
+      const labelResults = await gm.users.labels.list({
+        userId: emailAddress,
+      });
+      const doneLabelObj = labelResults.data.labels.filter(label => {
+        return label.name === 'SQRDone';
+      })[0];
+
+      const watchResults = await gm.users.watch({
+        userId: 'me',
+        requestBody: {
+          labelIds: [doneLabelObj.id],
+          topicName: 'projects/audio-seva-team-test/topics/gmail-labeled-done',
+        },
+      });
+
+      if (watchResults.data && doneLabelObj) {
+        // Store the most recent historyId
+        await storeHistoryIdInDatabase(
+          emailAddress,
+          watchResults.data.historyId
+        );
+        return res
+          .status(200)
+          .send(
+            `Watch (${watchResults.data.historyId}) created on ${
+              doneLabelObj.name
+            }/${doneLabelObj.id} for ${emailAddress}`
+          );
+      } else {
+        return res
+          .status(301)
+          .send(`Something went wrong, check logs for Gmail-initWatch`);
+      }
+    } catch (err) {
+      console.log('Error: ', err);
+      return res
+        .status(301)
+        .send(`Something went wrong, check logs for Gmail-initWatch`);
+    }
+  }
+);
+
+export const gmailDoneHandler = functions.pubsub
+  .topic('gmail-labeled-done')
+  .onPublish(async (message: Message) => {
+    const decodedMessage = JSON.parse(
+      Buffer.from(message.data, 'base64').toString('ascii')
+    );
+
+    // Gmail client and authenticate
+    const { client_key, secret, redirect } = functions.config().gmail;
+    const oauth2Client = new Google.google.auth.OAuth2(
+      client_key,
+      secret,
+      redirect
+    );
+    const { oauth, lastSyncHistoryId } = await fetchToken(
+      decodedMessage.emailAddress
+    );
+    oauth2Client.setCredentials({
+      access_token: oauth.token,
+      refresh_token: oauth.refreshToken,
+    });
+    const gm = await Google.google.gmail({
+      version: 'v1',
+      auth: oauth2Client,
+    });
+
+    const labelResults = await gm.users.labels.list({
+      userId: 'me',
+    });
+    const doneLabelObj = labelResults.data.labels.filter(label => {
+      return label.name === 'SQRDone';
```

Comment:
Again.
BTW, maybe we can store this id in the database?

Reply:
Any answer here?

Reply:
You mean the label id in the database? And if the id changes.

Reply:
Changed to "Done"

Reply:
>You mean the label id in the database? And if the id changes.
Yes, having ID in the database. I don't think it can change. However let's keep it as it is for now.